### ISSUE TITLE
Purity of numerical algorithms

### ISSUE THREAD

**Issue by berquist** [algorithm-archivists/algorithm-archive#510]
We had a discussion in https://github.com/algorithm-archivists/algorithm-archive/pull/495 about the purity of functions. The argument is that for many algorithms, specifically numerical ones, the implementation should not be pure for performance reasons, specifically memory. While the goal is not to write for performance, an algorithm should be written idiomatically for both 1. the language and (potentially) 2. how it is used. Also, from my experience, while it is easier to reason about the output of a pure numerical function by definition, it is not always easier to read. For algorithms of the scale in this book, this is probably a non-issue.

My proposal was to write all algorithms as being pure, but have text somewhere explaining the difference in the context of numerical algorithms, because it is an important topic not often touched upon.

**Comment by jiegillet** [comment]
+1. I suggest adding something in the [how to contribute](https://www.algorithm-archive.org/contents/how_to_contribute/how_to_contribute.html) page. Maybe add a point after style named purity and explain the reason why we use pure functions here.

**Comment by leios** [comment]
I think it fits best in the [Code style guide](https://www.youtube.com/watch?v=ijktorPAHGs) as a general principle.

I also agree that we should create a "disclaimer" chapter in place of the current how to contribute chapter (since the functionality of that one has been replaced by the how to contribute wiki). Basically, it would be nice to have some text describing the purpose of the code in the AAA.

**Comment by Butt4cak3** [comment]
I'm the one who originally pointed out that the current Thomas algorithm implementations aren't pure and I suggested we'd change that. The more I thought about it, the less convinced I became, though. The reason is that it adds a certain degree of complexity to the algorithms and makes it a bit more difficult to understand it by reading the code. The code is not as self-explanatory anymore.

So one thing we could do to counteract this is to add some comments in the reference (Julia) code and *require* other implementations to also include them. I'm thinking of something like what I did in my Thomas algorithm cleanup: https://github.com/algorithm-archivists/algorithm-archive/blob/9ca87e1af2c01ef314361a0c4780d0409fdaa2c4/contents/thomas_algorithm/code/java/Thomas.java#L4

**Comment by Yordrar** [comment]
I think that the goal of the code sample should be readability over performance, as there are a lot of possible ways to optimize code with many of them having some sort of obfuscation. The code should be as readable as possible so that anyone can understand it easily and adapt it to their needs. Now if that means that some functions should be pure and some of them shouldn't, so be it.

**Comment by leios** [comment]
My understanding here is that we would still like pure functions, but we also more heavily comment these functions to make it more clear what we are doing?

**Comment by Yordrar** [comment]
@leios well it depends on the algorithm, we can't really say for sure, maybe one is more readable but another one introduces a couple more steps or extra variables

**Comment by berquist** [comment]
Yes, I think that is a good approach (commenting + adding to how to contribute).
